contributors/design-proposals/api-machinery/udp-portforwarding.md

```diff
@@ -0,0 +1,109 @@
+# UDP Port Forwarding in Kubernetes
+
+Author: @shiywang
+
+## Abstract
+
+This describes an approach for providing support for:
+
+- port forwarding to containers with UDP protocol
+
+## Background && Motivation
+
+Fixes user request of https://github.com/kubernetes/kubernetes/issues/47862
+
+
+## Constraints and Assumptions
+
+- We already have an implemention with SPDY for TCP, the UDP implement basically reuse the SPDY framework, so it would be UDP through SPDY, the change here is remove the stream copy since UDP doesn't not has the same stream attribute as SPDY/TCP, so UDP will use datagram and `write` and `send` to communicate, socat will use unix domain socket as a middle layer to cross the boundaries between pod and Kubelet. and since TCP/SPDY doesn't have a message boundaries but UDP does, we also need to design an simple protocol for how to divide message read from SPDY back and forth.
```

Comment:
Please break into multiple lines if you don't mind.

Here are some edits to split up long sentences and make it easier to understand. Feel free to accept/reject as you see fit.

We already have a port forwarding implementation for TCP using SPDY. TCP is connection-based and supports streaming, so it's easy to identify individual connections and we can use `io.Copy` to transfer the streams of data between all the components involved (local client, kubectl, apiserver, kubelet/CRI, pod). Unlike TCP, UDP is connectionless and it uses messages instead of a stream of data. To add support for UDP, we will continue to use SPDY, but we will not be able to use `io.Copy` for UDP datagram transmission. Instead, we will design a simple protocol to allow UDP message encapsulation over SPDY. Additionally, we will use a Unix domain socket as the means for transferring UPD messages between the kubelet/CRI and the pod.

Reply:
thank you, this is much better : ) will update it.